proposals/custom-equals-in-value-classes.md

```diff
@@ -0,0 +1,262 @@
+# Custom equals in value classes
+
+* **Type**: Design proposal
+* **Authors**: Vladislav Grechko
+* **Status**: Prototype implemented
+* **Issue**: [KT-24874](https://youtrack.jetbrains.com/issue/KT-24874/)
+
+## Summary
+
+Allow overriding `equals` from `Any` and declaring so-called *typed equals* in value classes.
+
+## Motivation and use cases
+
+### Collections
+
+Suppose we want to implement wrapper for an object that implements `List` interface. We want this wrapper to be
+light-weight and avoid allocation of wrapper object. This might be achieved by declaring following inline class:
+
+```Kotlin
+@JvmInline
+value class SingletonList<T>(val element: T) : List<T> {
+    ...
+}
+```
+
+Since different implementations of `List` must be equal if they contain same elements, we have to
+customize `equals(other: Any?)` in `SingletonList`.
+
+> Note that passing instance of `SingletonList` as argument to non-inline function that takes `List`  will lead to
+> boxing and allocation a new object. For inline functions, boxing can be avoided when IR inliner is implemented.
+> Related issue:
+> [KT-40391](https://youtrack.jetbrains.com/issue/KT-40391/Inline-function-boxes-inline-class-type-argument)
+>
+
+### Units of measurement
+
+Inline classes are well suited to represent units of measurement. It might be useful to establish custom equivalence
+relation for them.
+
+```Kotlin
+@JvmInline
+value class Degrees(val value: Double)
+
+
+fun foo() {
+    // customize equality relation to make the set contain a single value
+    val uniqueAngles = setOf(Degrees(90.0), Degrees(-270.0))
+}
+```
+
+```Kotlin
+@JvmInline
+value class Minutes(val value: Int)
+
+@JvmInline
+value class Seconds(val value: Int)
+
+
+fun foo() {
+    // customize equality relation to make the set contain a single value
+    val uniqueTimeIntervals = setOf(Minutes(1), Seconds(60))
+}
+```
+
+## Proposal
+
+### Typed equals
+
+Simply overriding `equals(other: Any?)` in value class would require boxing of the right-hand side operand of
+every `'=='` comparison:
+
+```Kotlin
+@JvmInline
+value class Degrees(val value: Double) {
+    override fun equals(other: Any?): Boolean {
+        ...
+    }
+}
+
+fun foo() {
+    // have to box to pass to equals(other: Any?)
+    println(Degrees(0) == Degrees(45))
+}
+```
+
+That is why in [KEEP for inline classes](https://github.com/Kotlin/KEEP/blob/master/proposals/inline-classes.md) a new
+concept of **typed equals** was proposed:
+
+```Kotlin
+@JvmInline
+value class Degrees(val value: Double) {
+    fun equals(other: Degrees) = (value - other.value) % 360.0 == 0.0
+}
+```
+
+Type of `other` will be erased during compilation and passing argument will not require boxing.
+
+More precise, we define typed equals as a function such that:
+
+* Has name `"equals"`
```

Comment:
Better to make examples conform the restrictions. For example, the one above does not have the annotation.